File: BHAVRP_Python/cujae/inf/citi/om/assignment/classical/urgency/simplified.py
import numpy as np
from typing import List
from .by_urgency import ByUrgency
from .i_urgency import IUrgency
from ....service.osrm_service import OSRMService
from ....problem.input.problem import Problem
from ....problem.input.customer import Customer
from ....problem.input.depot import Depot
from ....problem.solution.solution import Solution
from ....problem.solution.cluster import Cluster

class Simplified(ByUrgency, IUrgency):
    
    def __init__(self):
        super().__init__()
        self.solution = Solution()
        self.list_clusters: List[Cluster]
        self.list_customers_to_assign: List[Customer]
        self.list_id_depots: List[int]
        
    def to_clustering(self):
        self.initialize()
        self.assign()
        return self.finish()
        
    def initialize(self):
        self.list_clusters = self.initialize_clusters()
        self.list_customers_to_assign = list(Problem.get_problem().get_customers())
        self.list_id_depots = list(Problem.get_problem().get_list_id_depots())
        
    def assign(self):
        urgency_matrix: np.ndarray = self.initialize_cost_matrix(self.list_customers_to_assign, Problem.get_problem().get_depots(), self.distance_type)
        closest_matrix: np.ndarray = self.initialize_cost_matrix(self.list_customers_to_assign, Problem.get_problem().get_depots(), self.distance_type)
        
        self.list_depots_ordered = self.get_depots_ordered(self.list_customers_to_assign, self.list_id_depots, closest_matrix)
        self.list_urgencies = self.get_list_urgencies(self.list_customers_to_assign, self.list_depots_ordered, urgency_matrix)

        while any(self.list_customers_to_assign) and self.list_clusters:
            pos_customer = self.list_urgencies.index(max(self.list_urgencies))
            customer = Problem.get_problem().get_customers()[pos_customer]
            id_customer = customer.get_id_customer()
            request_customer = customer.get_request_customer()

            id_closest_depot = self.list_depots_ordered[pos_customer][0]
            selected_depot: Depot = Problem.get_problem().get_depot_by_id_depot(id_closest_depot)
            capacity_depot = Problem.get_problem().get_total_capacity_by_depot(selected_depot)
            
            pos_cluster = self.find_cluster(id_closest_depot, self.list_clusters)
            
            if pos_cluster != -1:
                cluster: Cluster = self.list_clusters[pos_cluster]
                request_cluster = cluster.get_request_cluster()
                
                if capacity_depot >= (request_cluster + request_customer):
                    cluster.set_request_cluster(request_cluster + request_customer)
                    cluster.get_items_of_cluster().append(id_customer)

                    capacity_depot -= request_customer
                    
                    self.list_customers_to_assign[pos_customer] = None
                    self.list_urgencies[pos_customer] = 0.0
                    self.list_depots_ordered[pos_customer] = None
                else:
                    if capacity_depot > request_cluster:
                        self.list_depots_ordered[pos_customer].pop(0)
                        
                        if not self.list_depots_ordered[pos_customer]:
                            self.solution.get_unassigned_items().append(id_customer)
                            
                            self.list_customers_to_assign[pos_customer] = None
                            self.list_urgencies[pos_customer] = 0.0
                            self.list_depots_ordered[pos_customer] = None
                        else:
                            self.list_urgencies[pos_customer] = self.get_urgency(id_customer, self.list_depots_ordered[pos_customer], urgency_matrix)
                    else:
                        pos_depot = Problem.get_problem().find_pos_element(id_customer, self.list_id_depots[0], urgency_matrix)
                        
                        for i, depots_ordered in enumerate(self.list_depots_ordered):
                            if id_closest_depot in depots_ordered:
                                depots_ordered.remove(id_closest_depot)
                                if not depots_ordered:
                                    self.solution.get_unassigned_items().append(self.list_customers_to_assign[i].get_id_customer())
                                    
                                    self.list_customers_to_assign[i] = None
                                    self.list_urgencies[i] = 0.0
                                    self.list_depots_ordered[i] = None
                        
                        self.list_id_depots.pop(pos_depot)
                        
                        if self.list_clusters[pos_cluster].get_items_of_cluster():
                            self.solution.get_clusters().append(self.list_clusters.pop(pos_cluster))
                        else:
                            self.list_clusters.pop(pos_cluster)

    def finish(self):
        if self.list_customers_to_assign:
            for customer in self.list_customers_to_assign:
                if customer is not None:
                    self.solution.get_unassigned_items().append(customer.get_id_customer())
            
        if self.list_clusters:
            for cluster in self.list_clusters:
                if cluster.get_items_of_cluster():
                    self.solution.get_clusters().append(cluster)
                        
        OSRMService.clear_distance_cache()
        
        return self.solution
        
        # Método que retorna un listado con las urgencias de los clientes del listado entrado por parámetro.
    def get_list_urgencies(self, list_customers_to_assign: List[Customer], list_id_depots: List[List[int]], urgency_matrix: np.ndarray) -> List[float]:
        urgencies: List[float] = []

        if len(list_id_depots) > 1:
            for i, customer in enumerate(list_customers_to_assign):
                urgencies.append(self.get_urgency(customer.get_id_customer(), list_id_depots[i], urgency_matrix))
        else:
            for i, customer in enumerate(list_customers_to_assign):
                urgencies.append(self.get_urgency(customer.get_id_customer(), list_id_depots[0], urgency_matrix))
        return urgencies
    
    # Implementacion del método encargado de obtener la urgencia.
    def get_urgency(self, id_customer: int, list_id_depots: List[int], urgency_matrix: np.ndarray) -> float:
        urgency: float = 0.0
        closest_dist: float = 0.0
        other_dist: float = 0.0

        total_customers = Problem.get_problem().get_total_customers() + 1
        pos_matrix_customer: int = Problem.get_problem().get_pos_element(id_customer)
        depot_indices = [depot_id for depot_id in list_id_depots]
        
        closest_index = np.argmin(urgency_matrix[:, pos_matrix_customer])
        closest_dist = urgency_matrix[((depot_indices[closest_index]) - total_customers), pos_matrix_customer]
        
        urgency_matrix[((depot_indices[closest_index]) - total_customers), pos_matrix_customer] = float('inf')
        
        other_index = np.argmin(urgency_matrix[:, pos_matrix_customer])
        other_dist = urgency_matrix[((depot_indices[other_index]) - total_customers), pos_matrix_customer]
        
        if other_dist == float('inf'):
            urgency = closest_dist
        else:
            urgency = self.calculate_urgency(closest_dist, other_dist)

        return urgency

File: BHAVRP_Java/BHAVRPv2.0/BHAVRP2.0/src/cujae/inf/ic/om/assignment/classical/urgency/Simplified.java
package cujae.inf.ic.om.assignment.classical.urgency;

import java.util.ArrayList;
import java.util.Collections;

import cujae.inf.ic.om.problem.input.Customer;
import cujae.inf.ic.om.problem.input.Problem;

import cujae.inf.ic.om.problem.output.Cluster;
import cujae.inf.ic.om.problem.output.Solution;
import cujae.inf.ic.om.service.OSRMService;

import cujae.inf.ic.om.matrix.NumericMatrix;
import cujae.inf.ic.om.matrix.RowCol;

public class Simplified extends ByUrgency implements IUrgency {
	private Solution solution = new Solution();	
	
	private ArrayList<Cluster> listClusters;
	private ArrayList<Customer> listCustomersToAssign;

	private NumericMatrix urgencyMatrix;
	private NumericMatrix closestMatrix;
	
	private ArrayList<ArrayList<Integer>> listIDDepots;
	private ArrayList<ArrayList<Integer>> listDepotsOrdered;
	private ArrayList<Double> listUrgencies;
	
	public Simplified() {
		super();
	}
	
	@Override
	public Solution toClustering(){
		initialize();
		assign();
		return finish();
	}	
	
	@Override
	public void initialize() {
		listClusters = initializeClusters();
		listCustomersToAssign = new ArrayList<Customer>(Problem.getProblem().getCustomers());
		
		urgencyMatrix = initializeCostMatrix(listCustomersToAssign, Problem.getProblem().getDepots(), distanceType);
		closestMatrix = new NumericMatrix(urgencyMatrix);
		
		listIDDepots = new ArrayList<ArrayList<Integer>>();
		listIDDepots.add(Problem.getProblem().getListIDDepots());
		
		listDepotsOrdered = getDepotsOrdered(listCustomersToAssign, listIDDepots.get(0), closestMatrix);
		listUrgencies = getListUrgencies(listCustomersToAssign, listIDDepots, urgencyMatrix);
	}
	
	@Override
	public void assign() {
		int posCustomer = -1;
		int idCustomer = -1;
		double requestCustomer = 0.0;
		
		int idClosestDepot = -1;
		double capacityDepot = 0.0;
		
		int posCluster = -1;
		double requestCluster = 0.0;
		
		int totalItems = listCustomersToAssign.size();

		while((!listCustomersToAssign.isEmpty()) && (!listClusters.isEmpty())) 
		{
			posCustomer = listUrgencies.indexOf(Collections.max(listUrgencies));
			idCustomer = listCustomersToAssign.get(posCustomer).getIDCustomer();
			requestCustomer = listCustomersToAssign.get(posCustomer).getRequestCustomer();

			idClosestDepot = listDepotsOrdered.get(posCustomer).get(0).intValue();
			capacityDepot = Problem.getProblem().getTotalCapacityByDepot(Problem.getProblem().getDepotByIDDepot(idClosestDepot));

			posCluster = findCluster(idClosestDepot, listClusters);

			if(posCluster != -1)
			{
				requestCluster = listClusters.get(posCluster).getRequestCluster();
			
				if(capacityDepot >= (requestCluster + requestCustomer))
				{   
					requestCluster += requestCustomer;
					
					listClusters.get(posCluster).setRequestCluster(requestCluster);
					listClusters.get(posCluster).getItemsOfCluster().add(idCustomer);

					listCustomersToAssign.remove(posCustomer);
					listUrgencies.remove(posCustomer);
					listDepotsOrdered.remove(posCustomer);
				}
				else
				{
					if(capacityDepot > requestCluster)
					{	
						listDepotsOrdered.get(posCustomer).remove(0);
						
						if(listDepotsOrdered.get(posCustomer).isEmpty())
						{
							solution.getUnassignedItems().add(idCustomer);

							listCustomersToAssign.remove(posCustomer);
							listUrgencies.remove(posCustomer);
							listDepotsOrdered.remove(posCustomer);
						}
						else
							listUrgencies.set(posCustomer, getUrgency(idCustomer, listIDDepots.get(0), urgencyMatrix));
					}
					else
					{							
						int posDepot = Problem.getProblem().findPosElement(listIDDepots.get(0), idClosestDepot);
						int currentPosDepot = -1;
						int totalAvailableDepots = 0;
						
						for(int i = 0; i < listDepotsOrdered.size(); i++)
						{ 
							currentPosDepot = Problem.getProblem().findPosElement(listDepotsOrdered.get(i), idClosestDepot);

							if(currentPosDepot != -1)
							{
								totalAvailableDepots = listDepotsOrdered.get(i).size();

								if(totalAvailableDepots == 1)
								{
									solution.getUnassignedItems().add(listCustomersToAssign.get(i).getIDCustomer());
									
									listCustomersToAssign.remove(i);
									listUrgencies.remove(i);
									listDepotsOrdered.remove(i);
								}
								else
								{
									if(currentPosDepot == 0)
										listUrgencies.set(i, getUrgency(listCustomersToAssign.get(i).getIDCustomer(), listIDDepots.get(0), urgencyMatrix));
									else 
									{
										if(currentPosDepot == 1)
										{
											double currentUrgency = listUrgencies.get(i); 
											int posElement = Problem.getProblem().getPosElement(listCustomersToAssign.get(i).getIDCustomer());

											RowCol rcClosestDepot = urgencyMatrix.indexLowerValue(totalItems, posElement, (totalItems + listIDDepots.size() - 1), posElement);
											double secondDist = urgencyMatrix.getItem(rcClosestDepot.getRow(), rcClosestDepot.getCol());	
											urgencyMatrix.setItem(rcClosestDepot.getRow(), rcClosestDepot.getCol(), Double.POSITIVE_INFINITY);

											rcClosestDepot = urgencyMatrix.indexLowerValue(totalItems, posElement, (totalItems + listIDDepots.size() - 1),  posElement);
											double thirdDist = urgencyMatrix.getItem(rcClosestDepot.getRow(), rcClosestDepot.getCol());

											double firstDist = currentUrgency - secondDist;
											currentUrgency = calculateUrgency(firstDist, thirdDist);

											listUrgencies.set(i, currentUrgency);
										}
									}
									listDepotsOrdered.get(i).remove(currentPosDepot);
								}									
							}
						}
						listIDDepots.get(0).remove(posDepot);	
						
						if(!(listClusters.get(posCluster).getItemsOfCluster().isEmpty()))
							solution.getClusters().add(listClusters.remove(posCluster));
						else
							listClusters.remove(posCluster);
					}
				}
			}
		}
	}
	
	@Override
	public Solution finish() {
		if(!listCustomersToAssign.isEmpty())					
			for(int j = 0; j < listCustomersToAssign.size(); j++)	
				solution.getUnassignedItems().add(listCustomersToAssign.get(j).getIDCustomer());

		if(!listClusters.isEmpty())
			for(int k = 0; k < listClusters.size(); k++)
				if(!(listClusters.get(k).getItemsOfCluster().isEmpty()))
					solution.getClusters().add(listClusters.get(k));
		
		OSRMService.clearDistanceCache();
		
		return solution;
	}
	
	 /**
     * @param  ArrayList<Customer> listado de clientes
     * @param  ArrayList<Integer> listado de identificadores de depósitos
     * @param  NumericMatrix matriz con las distancias
     * @param  int identificador del depósito de mayor demanda insatisfecha
     * @return ArrayList<Double> listado de urgencia
     * Retorna un listado con las urgencias de los clientes del listado entrado por parámetro
     **/
	@Override
	public ArrayList<Double> getListUrgencies(ArrayList<Customer> listCustomersToAssign, ArrayList<ArrayList<Integer>> listIDDepots, NumericMatrix urgencyMatrix){
		ArrayList<Double> urgencies = new ArrayList<Double>();
		
		if(listIDDepots.size() > 1)
			for(int i = 0; i < listCustomersToAssign.size(); i++)
				urgencies.add(getUrgency(listCustomersToAssign.get(i).getIDCustomer(), listIDDepots.get(i), urgencyMatrix));
		else
			for(int i = 0; i < listCustomersToAssign.size(); i++)
				urgencies.add(getUrgency(listCustomersToAssign.get(i).getIDCustomer(), listIDDepots.get(0), urgencyMatrix));

		return urgencies;
	}

	/**
     * @param  int identificador del cliente
     * @param  ArrayList<Integer> listado de identificadores de depósitos
     * @param  NumericMatrix matriz con las distancias
     * @param int -1 identificador del depósito con mayor demanda insatisfecha
     * @return double urgencia del cliente
     * Retorna la urgencia del cliente 
     **/
	@Override
	public double getUrgency(int idCustomer, ArrayList<Integer> listIDDepots, NumericMatrix urgencyMatrix){
		double urgency = 0.0;
		double closestDist = 0.0;
		double otherDist = 0.0;
		RowCol rcClosetDepot = new RowCol();
		int posMatrixCustomer = -1;

		int totalCustomers = Problem.getProblem().getCustomers().size();
		posMatrixCustomer = Problem.getProblem().getPosElement(idCustomer);
			
		rcClosetDepot = urgencyMatrix.indexLowerValue(totalCustomers, posMatrixCustomer, (totalCustomers + listIDDepots.size() - 1), posMatrixCustomer);
		closestDist = urgencyMatrix.getItem(rcClosetDepot.getRow(), rcClosetDepot.getCol());	
		urgencyMatrix.setItem(rcClosetDepot.getRow(), rcClosetDepot.getCol(), Double.POSITIVE_INFINITY);

		rcClosetDepot = urgencyMatrix.indexLowerValue(totalCustomers, posMatrixCustomer, (totalCustomers + listIDDepots.size() - 1),  posMatrixCustomer);
		otherDist = urgencyMatrix.getItem(rcClosetDepot.getRow(), rcClosetDepot.getCol());
		
		if(otherDist == Double.POSITIVE_INFINITY)
			urgency = closestDist;
		else
			urgency = calculateUrgency(closestDist, otherDist);

		return urgency;
	}
}